### ISSUE TITLE
Union explicit array of values and reference to data

### ISSUE THREAD

**Issue by domoritz** [vega/vega#722]
Vega scales allow unioning of domains. However, it is not possible to union an explicit data domain and a reference to a data field. As a workaround, I can create a new data source with the values. It would be great if that was not necessary. 

I would like to write this:

```json
"domain": {
  "fields": [
    {"data": "source", "field": "foo"},
    [1,2,3,4]
  ]
}
```

**Comment by jheer** [comment]
Yes, that would be a useful addition. I'll look into it!

**Comment by jheer** [comment]
I looked into this and it seems the cleanest way to achieve this across all scale types is to generate an "anonymous" data set behind the scenes and populate it. One nice aspect is that we would support use of array-valued signals in addition to hard-wired array values. Basically, we would do what you are already doing manually, but manage it all "behind-the-scenes".

I can move forward with this feature, but I first would like to hear more about the use case(s) you envision. I just want to make sure this is of sufficient value to warrant the modifications. Thanks!

**Comment by domoritz** [comment]
We automatically try to union scale domains when we layer two views. Vega-Lite currently throws an error if a user specifies a domain for one layer as an array but not the other. If I could do what I described above, we could remove the error. 

I don't think there is a lot of value in supporting this, though. The use cases seem limited and we have bigger fish to fry.

**Comment by jheer** [comment]
Added in vega/vega-parser@8eef273b165fe32b4d77596c25f6f8d6cac1cc2a. Will be included in vega v3.0.0-beta.16. See below for an example test specification (use the console to inspect the scale state).

**Comment by jheer** [comment]
```json
{
  "signals": [
    {"name": "categories", "value": ["M", "N"]},
    {"name": "numbers", "value": [99, 100]}
  ],

  "data": [
    {
      "name": "source",
      "values": [
        {"a": "A", "b": 28},
        {"a": "B", "b": 55},
        {"a": "C", "b": 43},
        {"a": "D", "b": 91},
        {"a": "E", "b": 81},
        {"a": "F", "b": 53},
        {"a": "G", "b": 19},
        {"a": "H", "b": 87},
        {"a": "I", "b": 52}
      ]
    }
  ],

  "scales": [
    {
      "name": "linear",
      "type": "linear",
      "domain": {
        "fields": [
          {"data": "source", "field": "b"},
          [1, 2, 3],
          {"signal": "numbers"}
        ]
      },
      "range": [0, 1]
    },
    {
      "name": "quantile",
      "type": "quantile",
      "domain": {
        "fields": [
          {"data": "source", "field": "b"},
          [1, 2, 3],
          {"signal": "numbers"}
        ]
      },
      "range": [1, 2, 3, 4, 5]
    },
    {
      "name": "ordinal",
      "type": "ordinal",
      "domain": {
        "fields": [
          {"data": "source", "field": "a"},
          ["J", "K", "L"],
          {"signal": "categories"}
        ]
      },
      "range": ["a", "b", "c"]
    }
  ]
}
```

**Comment by domoritz** [comment]
It looks like there has been a regression for bin scales. 

```json
{
  "signals": [
    {"name": "categories","value": ["M","N"]},
    {"name": "bins","value": [1,2,3,4,5,6]}
  ],
  "scales": [
    {
      "name": "bin",
      "type": "bin-linear",
      "domain": {"signal": "bins"},
      "range": [0,1]
    }
  ]
}
```

```
ved.view._runtime.scales.bin.value.domain()
(6) [1, 2, 3, 4, 5, 6]
```

But unioning the same fields with themselves

```json
{
  "signals": [
    {"name": "categories","value": ["M","N"]},
    {"name": "bins","value": [1,2,3,4,5,6]}
  ],
  "scales": [
    {
      "name": "bin",
      "type": "bin-linear",
      "domain": {"fields": [{"signal": "bins"},{"signal": "bins"}]},
      "range": [0,1]
    }
  ]
}
```

```
ved.view._runtime.scales.bin.value.domain()
(2) [1, 6]
```

Of course, unioning domains for binned scales is super tricky but I wonder what your thoughts on this issue are. The use case is in Vega-Lite where two layers could use the same binned scale but different data. I realize that this is a constructed case and the binning is most likely not compatible if the data is different.

**Comment by jheer** [comment]
No regression here. Just a bug that was there all along :)

Fixed by vega/vega-parser@40c4aea7e5eb2ca0a3e9d927e0809929564cb267
